FIx trashed post show up in email listing and cannot be deleted

triple0t opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR addresses an issue where trashed email posts show up in the email listing and cannot be deleted. The changes include:

1. Added support for permanent deletion of email posts from the database (bypassing trash)
2. Implemented a new "Recreate email post" action in the email listing page
3. Added a new REST API endpoint for recreating email posts
4. Enhanced the trash modal UI to support both trash and permanent deletion modes


Note: The "Recreate email post" action will only appear when the transactional email does not have an associated post_id.


Closes #57550 WOOPLUG-4080

(For Bug Fixes) Bug introduced in PR # .

**Additional Documentation**

The PR introduces a new filter `woocommerce_email_editor_trash_modal_should_permanently_delete` in the email editor that controls whether emails should be permanently deleted or moved to trash. When set to `true`, emails will be permanently deleted instead of being moved to trash.


### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|    ![Screenshot 2025-04-29 at 10 11 19 AM](https://github.com/user-attachments/assets/d97e22da-3d78-481e-a183-b6d416c3286b)  |    ![Screenshot 2025-05-06 at 10 17 34 PM](https://github.com/user-attachments/assets/df50091f-dd21-41dc-8717-8f252d6ccc65)   |
| Current |  ![Screenshot 2025-05-29 at 6 46 52 PM](https://github.com/user-attachments/assets/08c75855-6549-46df-a7f8-27820a376816)  |
| Current | ![Screenshot 2025-05-29 at 6 46 56 PM](https://github.com/user-attachments/assets/593ffe53-ed03-4c78-915b-5efeb2c8be9b) |
| New |  ![Screenshot 2025-05-29 at 6 46 05 PM](https://github.com/user-attachments/assets/b6829759-c57f-4748-82fe-8562968b6710) |
| New | ![Screenshot 2025-05-29 at 6 46 12 PM](https://github.com/user-attachments/assets/b0c62f87-4e51-46af-9509-ce2324a613dd) |




<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Navigate to WooCommerce > Settings > Emails
2. Verify that emails without post IDs show a "Recreate email post" action
3. Click on "Recreate email post" and confirm that the post is recreated
4. Try to delete an email and verify:
   - The delete modal shows appropriate messaging for permanent deletion
   - The email is permanently deleted (skipping trash)
   - The email can be recreated from the listing page


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

triple0t committed:
Update README.md for email editor and add changelog for permanent deletion option in trash modal

* Improved formatting in the README for better readability.
* Added a new changelog entry detailing the support for a permanent deletion option in the trash modal, activated via a new filter.

triple0t committed:
Add filter for permanent email deletion in email editor integration

* Introduced a new filter to allow permanent deletion of email posts from the database, bypassing the trash. This enables users to recreate emails from the WooCommerce settings email listing page, enhancing email management functionality.

triple0t committed:
Add EmailListingRestController for managing email listings via the REST API

This will allow us perform AJAX actions from the new dataview email listing page

* Introduced EmailListingRestController to handle email listing operations via the WooCommerce REST API.
* Added nonce for email listing actions to enhance security.
* Updated WooCommerce class to register the new controller and its routes, improving email management capabilities.

triple0t committed:
Implement recreate email post functionality for Woo transactional emails in email listing

* Added a new `recreateEmailPost` function to the `useTransactionalEmails` hook, allowing users to recreate email posts via the WooCommerce settings email listing page.
* Integrated the `recreateEmailPost` function into the ListView component, enabling a new action for recreating email posts when the post ID is not present.
* Enhanced error handling for the email post recreation process, logging any issues encountered during the API call.

triple0t committed:
Fix linting errors

triple0t committed:
Add Woo changelog

triple0t committed:
Add trash email post functionality

* Introduced a new component for managing the trash action of email posts, allowing users to either permanently delete or move email posts to the trash.
* Implemented a modal for user confirmation with appropriate success and error notifications.
* Enhanced user experience by providing contextual messages based on the action taken (delete or move to trash).
* Integrated with existing WordPress data stores and hooks for seamless functionality.

triple0t committed:
Implement custom move to trash action for email posts

* Added a new function to modify the default move to trash action for email posts, replacing it with a custom action that integrates with the existing email editor functionality.
* Introduced a new file for the move-to-trash logic, enhancing the email management capabilities within the editor.
* Updated the blocks initialization to include the new move to trash action.

lysyjan commented on plugins/woocommerce/src/Internal/Admin/Emails/EmailListingRestController.php:
```diff
@@ -0,0 +1,195 @@
+	 */
+	private function get_args_for_recreate_email_post() {
+		return array(
+			'emailId'  => array(
```
I think the parameter `emailId` should be in the snakecase format for alignment with the rest of the codebase. So I would expect here `email_id`. What do you think?

triple0t replied:
Okay 👌🏿

triple0t replied:
Implemented in https://github.com/woocommerce/woocommerce/pull/57827/commits/47c1bdf62a6d1ee99118da850a03158bf30bbf9f

lysyjan reviewed (commented):
Great job @triple0t!

I have a couple of questions about the UI and compatibility:
1. What about the compatibility with the previous WP version? Is this feature fully compatible? Based on my testing, it looks like it isn't.

2. When I recreated the email post from the listing. It worked, but as a user, I was a bit confused because there wasn't any notification or message about the output. Should we improve it? Maybe in a follow-up ticket if needed.

3. Another confusion was related to the deleted email post. When I tried to edit an email notification, I expected a notice that the post for the specific email type had been deleted in the old email editor. For example, when one user deletes an email post and the second user wants to edit the same email later. He could be surprised by opening the old editor.

Here is an example how it could look like:
<img width="1545" alt="Screenshot 2025-06-02 at 9 26 45" src="https://github.com/user-attachments/assets/958dfb84-f5f1-4a10-9125-efe3a9ba770c" />

triple0t committed:
Add custom trash post support for WordPress 6.7 and 6.8.

`core.registerPostTypeSchema` is not available in WP 6.7, and `core.registerPostTypeActions` is not available in WP 6.8,
This commit aims to add backwards support for both versions of WP.

triple0t committed:
Refactor email ID parameter naming in API and frontend

* Updated the parameter name from `emailId` to `email_id` in both the REST controller and the frontend API call for consistency.
* Enhanced documentation in the EmailListingRestController to reflect the new naming convention, improving clarity for developers.

triple0t committed:
Prevent unnecessary initialization of methods

triple0t committed:
Fix REST API namespace conflict with EmailPreviewRestController

lysyjan reviewed (approved):
It works as expected with WP 6.8 and 6.7.

I tested deleting an email post for various email types and recreating it.
Here are some screenshots from my testing:
<img width="532" alt="Screenshot 2025-06-03 at 17 07 21" src="https://github.com/user-attachments/assets/817217c9-1dda-4216-9e87-cb8a454d2306" />
<img width="254" alt="Screenshot 2025-06-02 at 10 03 38" src="https://github.com/user-attachments/assets/93ed4e4c-e13c-4c26-a781-64aa8d45565c" />
